File: backend/test_recommender.py
import sys
import os
# Allow imports from backend/ folder
sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))

import unittest
import pandas as pd
from data_loader import RwandaDataLoader
from recommender import InziraRecommender

class TestInziraRecommender(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.loader = RwandaDataLoader()
        cls.recommender = InziraRecommender()
        cls.df = cls.loader.load_data()

    def test_dataset_loading_integrity(self):
        """Verify that the Rwanda places catalogue loads successfully with expected shapes and columns."""
        self.assertIsNotNone(self.df)
        self.assertGreater(len(self.df), 1000, "Dataset should contain at least 1,000 places.")
        required_columns = ["place_name", "latitude", "longitude", "category", "tags_list", "cost_level_norm", "duration_norm"]
        for col in required_columns:
            self.assertIn(col, self.df.columns, f"Required column '{col}' is missing.")

    def test_haversine_distance(self):
        """Verify mathematical correctness of the Haversine distance calculator."""
        # Great-circle distance from Kigali (Kigali Heights) to Gisenyi (Lake Kivu) is approx 96.5 km
        kigali_lat, kigali_lon = -1.9547, 30.0935
        gisenyi_lat, gisenyi_lon = -1.7019, 29.2625
        dist = self.recommender.haversine_distance(kigali_lat, kigali_lon, gisenyi_lat, gisenyi_lon)
        self.assertTrue(90.0 <= dist <= 110.0, f"Distance should be around ~96.5km, got {dist:.1f}km")

    def test_distance_score_tiers(self):
        """Verify the tiered distance scoring system maps correctly to distance limits."""
        self.assertEqual(self.recommender.calculate_distance_score(2.0), 100.0)    # <= 5km
        self.assertEqual(self.recommender.calculate_distance_score(12.0), 80.0)   # <= 15km
        self.assertEqual(self.recommender.calculate_distance_score(25.0), 60.0)   # <= 30km
        self.assertEqual(self.recommender.calculate_distance_score(50.0), 40.0)   # <= 60km
        self.assertEqual(self.recommender.calculate_distance_score(100.0), 25.0)  # <= 120km
        self.assertEqual(self.recommender.calculate_distance_score(150.0), 10.0)  # > 120km

    def test_preference_score_interests_expansion(self):
        """Verify that interest-to-tags expansion correctly matches places and assigns points."""
        place = pd.Series({
            "category": "Food & Drink",
            "tags_list": ["restaurant", "local", "rwandan"],
            "cost_level_norm": "medium",
            "duration_norm": "1-2 hours",
            "hidden_gem_score_norm": 50,
            "popularity_norm": "medium"
        })
        
        # Test case 1: Active 'food' interest gets expanded correctly and scores points
        prefs_with_food = {"interests": ["food"], "budget": "any", "available_time": "any", "hidden_gem_pref": "both"}
        score_food = self.recommender.calculate_preference_score(prefs_with_food, place)
        self.assertGreater(score_food, 50.0, "Interest matching 'food' should score preference match points.")

        # Test case 2: Unrelated interest evaluates to 0 due to strict filtering
        prefs_unrelated = {"interests": ["art"], "budget": "any", "available_time": "any", "hidden_gem_pref": "both"}
        score_unrelated = self.recommender.calculate_preference_score(prefs_unrelated, place)
        self.assertEqual(score_unrelated, 0.0, "Unrelated interests must yield 0 score for recommendations filtering.")

    def test_preference_score_budget_tiers(self):
        """Verify that budget matching filters budget categories correctly."""
        place_low_cost = pd.Series({
            "category": "Attraction",
            "tags_list": ["nature"],
            "cost_level_norm": "low",
            "duration_norm": "1-2 hours",
            "hidden_gem_score_norm": 50,
            "popularity_norm": "medium"
        })

        # Matches budget low
        prefs_low = {"interests": ["nature"], "budget": "low", "available_time": "any", "hidden_gem_pref": "both"}
        score_low = self.recommender.calculate_preference_score(prefs_low, place_low_cost)
        
        # Fails budget high (strictly compatible but not preferred)
        prefs_high = {"interests": ["nature"], "budget": "high", "available_time": "any", "hidden_gem_pref": "both"}
        score_high = self.recommender.calculate_preference_score(prefs_high, place_low_cost)
        
        self.assertGreater(score_low, score_high, "Budget matches should score higher than incompatible budgets.")

    def test_stage_1_main_recommendations(self):
        """Test Stage 1 recommendations retrieval & ordering."""
        prefs = {
            "interests": ["nature", "adventure"],
            "budget": "low",
            "available_time": "half-day",
            "preferred_province": "North",
            "hidden_gem_pref": "both"
        }
        recs = self.recommender.recommend_main_places(prefs, self.df, top_n=5)
        self.assertIsNotNone(recs)
        self.assertEqual(len(recs), 5, "Should return exactly top 5 recommendations.")
        # Check sorting
        scores = recs["final_score"].tolist()
        self.assertEqual(scores, sorted(scores, reverse=True), "Recommendations must be sorted descending by final score.")

    def test_stage_2_nearby_recommendations(self):
        """Test Stage 2 nearby recommendation logic including distance sorting & diversity bonus."""
        # Select a mock place (first entry in dataframe)
        selected_place = self.df.iloc[0]
        
        prefs = {
            "interests": ["food"],
            "budget": "any",
            "available_time": "any",
            "hidden_gem_pref": "both"
        }
        
        recs = self.recommender.recommend_nearby_places(selected_place, prefs, self.df, top_n=5)
        self.assertIsNotNone(recs)
        self.assertLess(len(recs), len(self.df))
        self.assertEqual(len(recs), 5)
        
        # Exclude selected place
        self.assertNotIn(selected_place["place_name"], recs["place_name"].tolist())
        # Verify distance calculations are attached
        self.assertIn("distance_km", recs.columns)
        self.assertLessEqual(recs.iloc[0]["distance_km"], recs.iloc[-1]["distance_km"] + 300, "Should score close places highly.")

if __name__ == "__main__":
    unittest.main()

File: lib/test_recommender.js
/**
 * test_recommender.js — Inzira
 * Command-line regression testing suite for Next.js recommendation logic.
 * Runs in standard Node.js to verify ES module algorithm calculations.
 */

const fs = require('fs');
const path = require('path');

// Simple assertion helper
function assert(condition, message) {
  if (!condition) {
    console.error(`❌ Assertion Failed: ${message}`);
    process.exit(1);
  }
}

(async () => {
  console.log('🧪 Starting JavaScript Recommendation Engine Tests...');

  // 1. Load the ES Module dynamically
  let recommender;
  try {
    recommender = await import('./recommender.js');
  } catch (err) {
    console.error('❌ Failed to import recommender.js. Ensure Node.js v18+ is used.', err);
    process.exit(1);
  }

  const {
    haversineDistance,
    calcPreferenceScore,
    calcDistanceScore,
    recommendMainPlaces,
    recommendNearbyPlaces
  } = recommender;

  // 2. Load the Places Catalogue
  let places;
  try {
    const filePath = path.join(__dirname, '../public/rwanda_places.json');
    places = JSON.parse(fs.readFileSync(filePath, 'utf8'));
    console.log(`✅ Dataset loaded successfully: ${places.length} places indexed.`);
    assert(places.length > 1000, 'Places index should contain >1000 places.');
  } catch (err) {
    console.error('❌ Failed to read or parse rwanda_places.json.', err);
    process.exit(1);
  }

  // 3. Test Haversine Distance
  const kigali = { lat: -1.9547, lon: 30.0935 };
  const gisenyi = { lat: -1.7019, lon: 29.2625 };
  const distance = haversineDistance(kigali.lat, kigali.lon, gisenyi.lat, gisenyi.lon);
  console.log(`✅ Haversine Distance (Kigali Heights -> Gisenyi): ${distance.toFixed(1)} km`);
  assert(distance >= 90.0 && distance <= 110.0, 'Distance must be roughly 90-110 km.');

  // 4. Test Distance Scores
  assert(calcDistanceScore(2) === 100, 'Distance score <= 5km must be 100');
  assert(calcDistanceScore(12) === 80, 'Distance score <= 15km must be 80');
  assert(calcDistanceScore(25) === 60, 'Distance score <= 30km must be 60');
  assert(calcDistanceScore(150) === 10, 'Distance score > 120km must be 10');
  console.log('✅ Distance score tiers verified successfully.');

  // 5. Test Preference Scoring & Interest Mappings
  const mockPlace = {
    place_name: 'Mock Cafe',
    category: 'Café',
    tags_list: ['coffee', 'relaxation'],
    cost_level_norm: 'medium',
    duration_norm: '1-2 hours',
    hidden_gem_score_norm: 80,
    popularity_norm: 'medium'
  };

  // Test Interest Match: food / relaxation should map to coffee/relaxation
  const prefsMatch = { interests: ['food'], budget: 'any', available_time: 'any', hidden_gem_pref: 'both' };
  const scoreMatch = calcPreferenceScore(prefsMatch, mockPlace);
  console.log(`✅ Preference Score (Matching food interest): ${scoreMatch} pts`);
  assert(scoreMatch > 50, 'Matching interests must score preference points.');

  // Test Interest Mismatch: shopping should evaluate to 0 since no tag overlaps
  const prefsMismatch = { interests: ['shopping'], budget: 'any', available_time: 'any', hidden_gem_pref: 'both' };
  const scoreMismatch = calcPreferenceScore(prefsMismatch, mockPlace);
  console.log(`✅ Preference Score (Mismatching shopping interest): ${scoreMismatch} pts`);
  assert(scoreMismatch === 0, 'Unrelated interests must score 0 and be filtered out.');

  // 6. Test Stage 1: Main Recommendations
  const userPrefs = {
    interests: ['nature'],
    budget: 'low',
    available_time: 'half-day',
    preferred_province: 'West',
    hidden_gem_pref: 'both'
  };
  const mainRecs = recommendMainPlaces(userPrefs, places, 5);
  console.log(`✅ Stage 1 main recommendations computed. Top pick: "${mainRecs[0].place_name}"`);
  assert(mainRecs.length === 5, 'Stage 1 should return exactly 5 recommendations.');
  assert(mainRecs[0].final_score >= mainRecs[4].final_score, 'Recommendations must be sorted descending.');

  // 7. Test Stage 2: Nearby recommendations
  const selectedPlace = mainRecs[0];
  const nearbyRecs = recommendNearbyPlaces(selectedPlace, userPrefs, places, 5);
  console.log(`✅ Stage 2 nearby recommendations computed around "${selectedPlace.place_name}".`);
  assert(nearbyRecs.length === 5, 'Stage 2 should return exactly 5 nearby picks.');
  assert(nearbyRecs.every(p => p.place_name !== selectedPlace.place_name), 'Selected place must be excluded from nearby picks.');

  console.log('\n🎉 All JavaScript Recommendation Scorer tests passed successfully!');
  process.exit(0);
})();
